### PR TITLE
port rotation_sources to C++

### PR DESCRIPTION
**Pull request by zingale**
## PR checklist

- [ ] test suite needs to be run on this PR
- [ ] this PR will change answers in the test suite to more than roundoff level
- [ ] all newly-added functions have docstrings as per the coding conventions
- [ ] the `CHANGES` file has been updated, if appropriate
- [ ] if appropriate, this change is described in the docs

### CODE REVIEW

**Comment by zingale** [comment]
tes results here: http://groot.astro.sunysb.edu/Castro/test-suite/gfortran/2020-06-21-007/index.html

File: Exec/science/wdmerger/Problem.H [hunk 1]
```diff
@@ -39,7 +39,6 @@ amrex::Real dot_product(const amrex::Real a[], const amrex::Real b[]);
 
 // Computes standard cross-product of two three-vectors.
```
Comment (maxpkatz) [inline Exec/science/wdmerger/Problem.H:40]: Can you please remove this comment too

File: Source/driver/math.H [hunk 2]
```diff
@@ -0,0 +1,14 @@
+#ifndef CASTRO_MATH_H
+#define CASTRO_MATH_H
+
+AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
```
Comment (maxpkatz) [inline Source/driver/math.H:4]: Let's drop back to `AMREX_INLINE`, I think it was a mistake that I let `AMREX_FORCE_INLINE` proliferate.

File: Source/rotation/Castro_rotation.H [hunk 3]
```diff
@@ -39,3 +39,78 @@
...
+                                 Array4<Real const> const& state);
+
+///
+/// Compute the rotational acceleration for a single zone
+/// (Coriolis and centrifugal)
+///
+/// @param r         distance from origin of rotation vector
+/// @param v         velocity
+/// @param omega     rotation vector
+/// @param corilois  do we include the Coriolis force
```
Comment (maxpkatz) [inline Source/rotation/Castro_rotation.H:110]: Spelling

File: Source/rotation/rotation_sources.cpp [hunk 4]
```diff
@@ -0,0 +1,489 @@
...
+    src[UMZ] = Sr[2];
+
+    snew[UMX] += dt * src[UMX];
+    snew[UMY] += dt * src[UMY];
+    snew[UMZ] += dt * src[UMZ];
+
+#ifdef HYBRID_MOMENTUM
+    if (state_in_rotating_frame == 1) {
+      GpuArray<Real, 3> linear_source;
+      linear_source[0] = src[UMR];
```
Comment (maxpkatz) [inline Source/rotation/rotation_sources.cpp:62]: In `set_hybrid_momentum_source`, the input `linear_source` is the source to the linear momenta UMX, UMY, and UMZ, and the output `hybrid_source` is the source to UMR, UML, and UMP. See construct_old_gravity_source. This is different from the interface in Fortran.

File: Source/rotation/rotation_sources.cpp [hunk 5]
```diff
@@ -0,0 +1,489 @@
...
+    snew[UMX] += dt * src[UMX];
+    snew[UMY] += dt * src[UMY];
+    snew[UMZ] += dt * src[UMZ];
+
+#ifdef HYBRID_MOMENTUM
+    // The source terms vanish if the state variables are measured in the
+    // inertial frame; see wdmerger paper III.
+
+    if (state_in_rotating_frame == 1) {
+      GpuArray<Real, 3> linear_source;
```
Comment (maxpkatz) [inline Source/rotation/rotation_sources.cpp:373]: Same comment about `set_hybrid_momentum_source`

File: Exec/science/wdmerger/Problem.H (hunk 1 above)
Comment (zingale) [reply Exec/science/wdmerger/Problem.H:40]: done

File: Source/driver/math.H (hunk 2 above)
Comment (zingale) [reply Source/driver/math.H]: done

File: Source/rotation/Castro_rotation.H (hunk 3 above)
Comment (zingale) [reply Source/rotation/Castro_rotation.H]: fixed

File: Source/rotation/rotation_sources.cpp (hunk 4 above)
Comment (zingale) [reply Source/rotation/rotation_sources.cpp]: I think I've fixed this now

**Pull request merged**